packages/contracts-periphery/contracts/universal/op-nft/CitizenshipBadge.sol

```diff
@@ -0,0 +1,57 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.15;
+
+import "@rari-capital/solmate/src/tokens/ERC721.sol";
+import "@openzeppelin/contracts/utils/Strings.sol";
+import "./CitizenshipChecker.sol";
+import "./SocialContract.sol";
+
+contract CitizenshipBadge is ERC721 {
+    address public immutable admin;
+
+    SocialContract public immutable socialContract;
+    CitizenshipChecker public immutable citizenshipChecker;
+
+    constructor(
+        address _admin,
+        address _socialContract,
+        address _citizenChecker
+    ) ERC721("Citizenship", "CB") {
```

Comment:
The ERC721 name and symbol are set in storage in the constructor. To deploy this behind a proxy, we will need to set them in an `initialize()`. I don't think there is a good way to do this without using assembly

Reply:
Basically need to know the storage layout of the ERC721 and use `sstore` to set the values `"Citizenship"` and `"CB"` in the correct slots. Wondering if you have any thoughts here @clabby

Reply:
Hey! Sorry for the delay here @tynes, this mention got buried in my notifications and saw it while cleaning them out tonight.

Because both of these strings are less than 32 bytes in length, this actually isn't too bad. Strings <= 31 bytes in length are encoded in one word with the contents padded to the left and the number of nibbles (`len(bytes) * 2`) in the last byte.

To store the `"Citizenship"` string:
```yul
assembly {
    // "Citizenship" string (11 bytes in length)
    let citizenshipStr := 0x436974697A656E73686970
    // Number of nibbles in the string (11 * 2 = 22)
    let sLen := 22
    // Change this accordingly
    let nameSlot := 0

    // Store the "Citizenship" string in `nameSlot`
    // Note: The `citizenshipStr` is shifted left by 168 (21 bytes) to align its bytes to the left
    // of the word. This value will change depending on the byte length of the string:
    // (32 - len(bytes)) * 8
    sstore(nameSlot, or(shl(168, citizenshipStr), sLen))
}
```

Alternatively, you could pre-format these words rather than doing it at runtime, but it does increase codesize:
```yul
assembly {
    // Change these accordingly
    let nameSlot := 0
    let symbolSlot := 1

    // Store "Citizenship" string in `nameSlot`
    sstore(nameSlot, 0x436974697a656e73686970000000000000000000000000000000000000000016)
    // Store "CB" string in `symbolSlot`
    sstore(symbolSlot, 0x4342000000000000000000000000000000000000000000000000000000000004)
}
```